### PR TITLE
llext: export DT devices to extensions

### PR DESCRIPTION
**Pull request by pillo79**
This PR adds support for automatic export of all Zephyr devices to extensions via a new Kconfig option, `CONFIG_LLEXT_EXPORT_DEVICES`. When this option is enabled, all Devicetree macros in `zephyr/device.h` can be used in extensions as well to inspect and interact with the devices configured in the current build.

Calling `EXPORT_SYMBOL` with anything more complicated than a simple identifier resulted in macro expansion issues. This has been fixed by adding a level of indirection, moving the actual symbol definition code to private `Z_` macros. A number of related comments have been updated. 

New test code has been added to properly verify the new support. The test retrieves the device object associated with the Zephyr console via `DT_CHOSEN` macros and verifies it matches with the one obtained at runtime by calling `device_get_binding()`. This check is currently unsupported on Xtensa since the `qemu_xtensa` platform defines no usable DT devices.    

Also fixes 2 minor issues in `device.h`:
- The error message for `Z_DEVICE_NAME_CHECK` included a call to `DEVICE_NAME_GET`, but the argument was already a string. This resulted in a confusing error message being displayed when the check failed.
- There was a misleading comment that implied the name returned by `Z_DEVICE_DT_DEV_ID` was used in `device_get_binding()` instead of the DT path, and must thus be limited to `Z_DEVICE_MAX_NAME_LEN` chars. 
This is not the case (anymore?): `device_get_binding()` matches neither paths nor identifiers, but [either DT _labels_ or DT _full names_](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/device.h#L152-L153) as returned by the `DEVICE_DT_NAME` macro [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/device.h#L199).

### CODE REVIEW

File: include/zephyr/device.h [hunk 1]
```diff
@@ -91,10 +91,8 @@ typedef int16_t device_handle_t;
  */
 #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
 
-/* Node paths can exceed the maximum size supported by
- * device_get_binding() in user mode; this macro synthesizes a unique
- * dev_id from a devicetree node while staying within this maximum
- * size.
+/* This macro synthesizes a unique dev_id from a devicetree node by using
+ * the node's dependency ordinal.
```
Comment (lyakh) [inline include/zephyr/device.h:96]: I think you're modifying a comment for `Z_DEVICE_DT_DEV_ID()` and not for `DEVICE_NAME_GET()` as the commit description seems to suggest?

Comment (pillo79) [reply include/zephyr/device.h:96]: Thanks for spotting, will fix the commit description along with the reviews :+1:

**Comment by fabiobaltieri** [comment]
@gmarull can you take a look?

File: include/zephyr/device.h [hunk 2]
```diff
@@ -999,7 +999,7 @@ device_get_dt_nodelabels(const struct device *dev)
  */
 #define Z_DEVICE_NAME_CHECK(name)                                              \
 	BUILD_ASSERT(sizeof(Z_STRINGIFY(name)) <= Z_DEVICE_MAX_NAME_LEN,       \
-			    Z_STRINGIFY(DEVICE_NAME_GET(name)) " too long")
+			    Z_STRINGIFY(name) " too long")
```
Comment (gmarull) [inline include/zephyr/device.h:1011]: ack

File: include/zephyr/device.h [hunk 3]
```diff
@@ -197,7 +207,8 @@ typedef int16_t device_handle_t;
 			DEVICE_DT_NAME(node_id), init_fn, pm, data, config,    \
 			level, prio, api,                                      \
 			&Z_DEVICE_STATE_NAME(Z_DEVICE_DT_DEV_ID(node_id)),     \
-			__VA_ARGS__)
+			__VA_ARGS__);                                          \
+	Z_DEVICE_EXPORT(node_id)
```
Comment (gmarull) [inline include/zephyr/device.h:211]: you can use `IF_ENABLED(CONFIG_LLEXT_EXPORT_DEVICES, (EXPORT_SYMBOL(DEVICE_DT_NAME_GET(node_id))))`

File: include/zephyr/device.h [hunk 4]
```diff
@@ -12,6 +12,7 @@
 #include <zephyr/devicetree.h>
 #include <zephyr/init.h>
 #include <zephyr/linker/sections.h>
+#include <zephyr/llext/symbol.h>
```
Comment (gmarull) [inline include/zephyr/device.h:15]: conditional include?

File: include/zephyr/device.h [hunk 5]
```diff
@@ -99,6 +100,14 @@ typedef int16_t device_handle_t;
...
 #define Z_DEVICE_DT_DEV_ID(node_id) _CONCAT(dts_ord_, DT_DEP_ORD(node_id))
 
+#if defined(CONFIG_LLEXT_EXPORT_DEVICES)
+/* Export device identifiers using the builtin name */
+#define Z_DEVICE_EXPORT(node_id) EXPORT_SYMBOL(DEVICE_DT_NAME_GET(node_id))
+#else
+/* Do not export device identifiers */
+#define Z_DEVICE_EXPORT(node_id)
+#endif
+
```
Comment (gmarull) [inline include/zephyr/device.h:110]: see my other comment

File: include/zephyr/device.h [hunk 6]
```diff
@@ -171,7 +180,8 @@ typedef int16_t device_handle_t;
  *
  * The device is declared with extern visibility, so a pointer to a global
  * device object can be obtained with `DEVICE_DT_GET(node_id)` from any source
- * file that includes `<zephyr/device.h>`. Before using the pointer, the
+ * file that includes `<zephyr/device.h>` (even from extensions, when @c
+ * CONFIG_LLEXT_EXPORT_DEVICES is enabled). Before using the pointer, the
```
Comment (gmarull) [inline include/zephyr/device.h:184]: wrong doxygen use, use `@kconfig{CONFIG_...}`

File: include/zephyr/device.h [hunk 7]
```diff
@@ -197,7 +207,8 @@ typedef int16_t device_handle_t;
 			DEVICE_DT_NAME(node_id), init_fn, pm, data, config,    \
 			level, prio, api,                                      \
 			&Z_DEVICE_STATE_NAME(Z_DEVICE_DT_DEV_ID(node_id)),     \
-			__VA_ARGS__)
+			__VA_ARGS__);                                          \
```
Comment (gmarull) [inline include/zephyr/device.h:210]: shouldn't trailing `;` be conditioned to the next statement? if not, move export above.

File: include/zephyr/device.h (hunk 3 above)
Comment (pillo79) [reply include/zephyr/device.h]: I could, but the separate macro is easier to extend in the (short) future; I have a draft of a more complex proposal (https://github.com/zephyrproject-rtos/zephyr/pull/77799) where there are two different implementations for `Z_DEVICE_EXPORT`. 
For this reason I'm proposing to keep the separate macro, but included within an `IF_ENABLED` in the parent.

**Comment by pillo79** [comment]
**v2**:
- new proposal for  `Z_DEVICE_EXPORT`
- implemented all other notes by @gmarull 
- renamed new private macros from `macro_IMPL` to `Z_macro` to keep in line with Zephyr internals

**Review by teburd — APPROVED** [review]

**Review by gmarull — APPROVED** [review]
ack device changes

**Pull request merged**